crypto/merklekeystore/persistentKeystore.go

```diff
@@ -0,0 +1,150 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package merklekeystore
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/util/db"
+)
+
+const keystoreSchemaVersion = 1
+const keystoreTableSchemaName = "merklekeystore"
+
+// PersistentKeystore Provides an abstraction to the keystore, the DB operations for fetching/storing them and an internal caching mechanism
+type PersistentKeystore struct {
+	store db.Accessor
+}
+
+// Persist dumps the keys into the database as separate row for each key
+func (p *PersistentKeystore) Persist(keys []crypto.SignatureAlgorithm, firstValid uint64, interval uint64) error {
+	if keys == nil {
+		return fmt.Errorf("no keys provided (nil)")
+	}
+
+	err := p.store.Atomic(func(ctx context.Context, tx *sql.Tx) error { // does it take too long?
+		err := keystoreInstallDatabase(tx) // assumes schema table already exists (created by partInstallDatabase)
+		if err != nil {
+			return err
+		}
+
+		round := indexToRound(firstValid, interval, 0)
+		for i, key := range keys {
+			_, err := tx.Exec("INSERT INTO BlockProofKeys (id, round, key) VALUES (?,?,?)", i, round, protocol.Encode(&key))
+			if err != nil {
+				return fmt.Errorf("failed to insert blockProof key number %v round %d. SQL Error: %w", i, round, err)
+			}
+			round += interval
+		}
+
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("PersistentKeystore.Persist: %w", err)
+	}
+
+	return nil // Success
+}
+
+// GetKey receives a round number and returns the corresponding (previously committed on) key.
+func (p *PersistentKeystore) GetKey(round uint64) (*crypto.SignatureAlgorithm, error) {
+	var keyB []byte
+	err := p.store.Atomic(func(ctx context.Context, tx *sql.Tx) error { // does it take too long?
```

Comment:
can remove this comment. + line after

Reply:
i'll remove it but we still didn't answer the question